### PR TITLE
[System.Data] move to corefx

### PR DESCRIPTION
**Pull request by EgorBo**
There were few major challenges:
1) System.Data in corefx is represented by 3 separate assemblies (System.Data.Common, System.Data.SqlClient and System.Data.Odbc) while mono has single System.Data assembly.
2) OLEDB provider is currently not supported (it was deprecated long time ago and windows-only so it doesn't look like .NET Core team is going to bring it back).
3) Since I had to fill some API gaps (like SqlDependency, etc) using old code there are a few helpers in System.Data/corefx folder.
4) Lots of changes were made in the upstream - https://github.com/dotnet/corefx/pulls?q=is%3Apr+author%3AEgorBo+is%3Aclosed (and [8 issues](https://github.com/dotnet/corefx/issues?utf8=%E2%9C%93&q=is%3Aissue%20author%3AEgorBo%20) were filed)

**API diffs** (minor issuses - will be fixed with the next PR to corefx):
SqlDataReader
- GetData (should be virtual)
- IsCommandBehavior (should not be internal)

SqlCommand (the following methods are private in corefx but should be public)
- BeginExecuteNonQuery
- BeginExecuteXmlReader
- EndExecuteNonQuery
- EndExecuteReader
- EndExecuteXmlReader

LinqDataView is missing (System.Data.DataSetExtensions assembly)

Other "red" lines in API Diff are System.Component.* & MonoTODO attributes. Also if a method was removed but still exists in its parent (same signature) API Diff will treat it like a breaking changes (but it actually is not).

## Tests
**Integration tests:** (MSSQL via SqlClient, MySQL via Odbc)
Passed: 304 , Failures: 0

**All tests:**
Tests run: 2407, Passed: 2351, Errors: 0, Ignored: 56

### CODE REVIEW

File: mcs/class/System.Data/corefx/RuntimeInformation.cs [hunk 1]
```diff
@@ -0,0 +1,29 @@
+using System;
```
Comment (marek-safar) [inline mcs/class/System.Data/corefx/RuntimeInformation.cs]: I still don't like this copy, if we you really need it we need to split Facades build so you can take dependency on Facades/System.Runtime.InteropServices.RuntimeInformation

**Comment by flutos** [comment]
I noticed this error related to interop certificates trying to build with this patch and was wondering if there was something I might be missing?

../build/library.make:185: warning: ignoring old recipe for target `test-local'
Makefile:25: warning: overriding recipe for target `test-local'
../../build/library.make:185: warning: ignoring old recipe for target `test-local'
make[8]: *** No rule to make target `../../../external/corefx/src/common/src/Interop/Windows/Crypt32/Interop.certificates.cs', needed by `../../class/lib/net_4_x-linux/System.Data.dll'.  Stop.
make[7]: *** [do-all] Error 2
make[6]: *** [all-recursive] Error 1

**Comment by EgorBo** [comment]
@flutos this PR is a Work-in-progress [WIP], currently it doesn't compile since it needs some corefx changes for example.

**Comment by flutos** [comment]
Thats ok , im eager to test the core sqlclient bits warts and all.  I got it all compiled now with your latest patch  but there is a problem I noticed referencing the xml file System.Data.SqlClient.SqlMetaData.xml which is coming back with a null xml stream . Im not sure if I may be missing something else :

    protected override DbMetaDataFactory CreateMetaDataFactory(DbConnectionInternal internalConnection, out bool cacheMetaDataFactory)
        {
            Debug.Assert(internalConnection != null, "internalConnection may not be null.");

            Stream xmlStream = System.Reflection.Assembly.GetExecutingAssembly().GetManifestResourceStream("System.Data.SqlClient.SqlMetaData.xml");
            cacheMetaDataFactory = true;

**Comment by EgorBo** [comment]
@flutos good catch! seems, the file is not added to the build as an embedded resource.

**Comment by borgdylan** [comment]
ping, is this still on the cards?

**Comment by EgorBo** [comment]
@borgdylan of course! will update it soon with my local changes (needs some corefx changes).

**Comment by marek-safar** [comment]
build

**Comment by EgorBo** [comment]
@monojenkins build

**Comment by flutos** [comment]
Just wanted to pass along my most recent tests of your latest changes. Everything built successfully so I then ran our full suite of unit tests and integrated tests. There are around 30,000 of them and it takes around 2 days to run them all but after running them all and hitting sqlserver constantly with this sqlclient driver everything worked perfectly without errors. So then I ran some performance tests to compare the sqlclient performance of it with sqlclient on windows and the old mono sqlclient driver. Attached are my results but I found it was a lot faster than the old mono driver and the bulk inserting worked great which is better than the crashing we currently get with the current mono sqlclient driver. Comparing  it to windows was quite positive too in that it was at least as fast as the windows .net sqlclient driver if not faster in other tests . Over all though its all looks very good.

thanks
scott
[a1.txt](https://github.com/mono/mono/files/1223348/a1.txt)

**Comment by EgorBo** [comment]
@flutos Thank you so much for your feedback! By "Windows sqlclient" do you mean .NET core or .NET 4.x?
there can be some small performance differences since .NET Core for Windows (as well as the old .net) use native transport implementation (sni.dll).

**Comment by flutos** [comment]
On windows we are using .net 4.6

**Comment by EgorBo** [comment]
@monojenkins build

**Comment by marek-safar** [comment]
commit apidiff

**Comment by akoeplinger** [comment]
👏

**Pull request merged**